### DISCUSSION TITLE
[v2.0] Remove redundancies: SinglePointerGesture - DualPointerGesture

### DISCUSSION THREAD

**Discussion by biodiv** [biodiv/contactjs#17] [General]
During the restructuring of the codebase, the follwing issue showed up. The classes `SinglePointerGesture` and `DualPointerGesture` are both subclasses of `Gesture`. `SinglePointerGesture` and `DualPointerGesture` have different gesture parameters: `SinglePointerGestureParameters` and `DualPointerGestureParameters`, which are used to recognize the gestures.

Both `SinglePointerGesture` and `DualPointerGesture` require a method of getting the appropriate parameter values from `SinglePointerInput` or `DualPointerInput`. The logic of reading those parameters is the same, but I did not find a way to respect the different types. The result are redundant methods:

For `DualPointerGesture`:
```
getPointerInputGlobalValue(
    pointerInput: DualPointerInput,
    parameterName: keyof DualPointerGestureGlobalParameters
  ): GestureParameterValue {
    const pointerInputValue = pointerInput.parameters.global[parameterName];
    return pointerInputValue;
}
```

and for `SinglePointerGesture`:
```
getPointerInputGlobalValue(
    pointerInput: SinglePointerInput,
    parameterName: keyof SinglePointerGestureGlobalParameters
): GestureParameterValue {
    const pointerInputValue = pointerInput.parameters.global[parameterName];
    return pointerInputValue;
}
```

I tried several things with TypeScript generics, but have not found a solution yet. Any input is highly appreciated.

**Comment by SE2Dev** [comment]
The parameters were one of the things where I wasn't entirely satisfied with when setting up the types. I wasn't really able to find a clean way to pass the parameter lists around in a way that was both clean and provided adequate static type safety.

If I understand this correctly, you want to move `getPointerInputGlobalValue` and `getPointerInputLiveValue` into the `Gesture` base class, right?

As it stands, I think you'd still need separate functions to explicitly state that `SinglePointerGesture` would need to accept a `SinglePointerInput` and that `DualPointerGesture` would need to accept a `DualPointerInput`. The actual implementation itself could exist on the base gesture class and accept a base `PointerInput` and `string`. Then the overrides in `SinglePointerGesture` and `DualPointerGesture` can simple call the parent method using `super`.

I'm not sure how much I like this solution though, as it would introduce a slight runtime performance penalty - which, when it comes to gestures and user interactions, I'm not particularly fond of.

**Comment by biodiv** [comment]
Yes, I wanted to move those methods to the `Gesture`class.

I changed the structure of the interfaces and replaced the `minMaxInterval` Arrays with the keys `.min`and `.max`. That way I could get rid of the `getPointerInputGlobalValue` and `getPointerInputLiveValue` methods.

However, the method `validateGestureParameters` still could not be moved to the `Gesture` class and had to exist on both `SinglePointerGesture` and `DualPointerGesture`. The interfaces `SinglePointerGestureParameters` and `DualPointerGestureParameters`, have different properties and could not be compared with the parameters of `(Dual)PointerInput.parameters`.

I converted instances being of the type  `SinglePointerGestureParameters`, `DualPointerGestureParameters` or `(Dual)PointerParameters`  to a `Record<>` type, which allowed me to compare them.

I am not yet happy with this solution, especially with the `any` in this line (Gesture.ts line 165):
`const timedPointerInputValues = pointerInput.parameters[timespan] as Record<string, any>;`

but I think it's still better than repeating the whole method in two subclasses.

Still, a better solution would be nice.
